src/bit-fields/src/lib.rs

```diff
@@ -0,0 +1,1557 @@
+                if x <= Self::MAX {
+                    // SAFETY:
+                    // See [`Self::data`].
+                    let cur = (Self::MASK & *self.0) >> START;
+                    if x <= Self::MAX - cur {
+                        let shift = x << START;
+                        *self.0 += shift;
+                        Ok(())
+                    } else {
+                        Err(CheckedAddAssignError::Overflow)
+                    }
+                } else {
+                    Err(CheckedAddAssignError::OutOfRange)
+                }
+            }
+
+            /// Subtract `x` from the value of the bit range.
+            ///
+            /// # Errors
+            ///
+            /// 1. When `x` is greater than the maximum value storable in the bit range.
+            /// 2. When subtracting `x` from the value of the bit range would underflow.
+            // `x <= Self::MAX` guarantees `x << START` is safe and `x <= cur` guarantees
+            // `self.data_mut() -= shift` is safe.
+            #[allow(clippy::integer_arithmetic, clippy::arithmetic)]
+            #[inline]
+            pub fn checked_sub_assign(&mut self, x: $x) -> Result<(), CheckedSubAssignError> {
+                if x <= Self::MAX {
+                    // SAFETY:
+                    // See [`Self::data`].
+                    let cur = (Self::MASK & *self.0) >> START;
+                    if x <= cur {
+                        let shift = x << START;
+                        *self.0 -= shift;
+                        Ok(())
+                    } else {
+                        Err(CheckedSubAssignError::Underflow)
+                    }
+                } else {
+                    Err(CheckedSubAssignError::OutOfRange)
+                }
+            }
+
+            /// Sets the bit range returning `Err(())` when the given `x` is not storable in the
+            /// range.
+            ///
+            /// # Errors
+            ///
+            /// When `x` is greater than the maximum storable value in `self`.
+            // `x <= Self::MAX` guarantees `x << START` is safe.
+            #[allow(clippy::integer_arithmetic, clippy::arithmetic)]
+            #[inline]
+            pub fn checked_assign(&mut self, x: $x) -> Result<(), CheckedAssignError> {
+                if x <= Self::MAX {
+                    let shift = x << START;
+                    *self.0 = shift | (*self.0 & !Self::MASK);
+                    Ok(())
+                } else {
+                    Err(CheckedAssignError)
+                }
+            }
+            /// Returns `Self::MAX`.
+            #[must_use]
+            #[inline]
+            pub const fn get_max(&self) -> $x {
+                Self::MAX
+            }
+            /// Returns `Self::MASK`.
+            #[must_use]
+            #[inline]
+            pub const fn get_mask(&self) -> $x {
+                Self::MASK
+            }
```

Comment:
I am assuming we don't use the constants directly because the type parameters make it very awkward? If so, maybe document that in the code

Reply:
Yes these are mostly just for convenience, it is easier to write `field_u128.two().get_max()` than `BitRange::<u128, 1, 3>::MAX`. It also helps in testing such that we can generate a bit field then test a specific field's size to essentially check it is the correct type (without this it is difficult to test a field of a generated bit field is the right size). I have added the comments `/// Conveniance alias for [`Self::MAX`].` and `/// Conveniance alias for [`Self::MASK`].` respectively.